evhtp.c

```diff
@@ -2196,55 +2249,66 @@ htp__connection_writecb_(struct bufferevent * bev, void * arg)
      * to make sure we are not over it. If we have gone over the max we set the
      * keepalive bit to 0, thus closing the connection.
      */
-    if (c->htp->max_keepalive_requests)
+    if (keepalive_max > 0)
     {
-        if (++c->num_requests >= c->htp->max_keepalive_requests)
+        conn->num_requests += 1;
+
+        if (conn->num_requests >= keepalive_max)
         {
-            HTP_FLAG_OFF(c->request, EVHTP_REQ_FLAG_KEEPALIVE);
+            HTP_FLAG_OFF(conn->request, EVHTP_REQ_FLAG_KEEPALIVE);
         }
     }
 
-    if (c->request->flags & EVHTP_REQ_FLAG_KEEPALIVE)
+    if (conn->request->flags & EVHTP_REQ_FLAG_KEEPALIVE)
     {
         htp_type type;
 
-        htp__request_free_(c->request);
+        /* free up the current request, set it to NULL, making
+         * way for the next request.
+         */
+        evhtp_safe_free(conn->request, htp__request_free_);
 
-        HTP_FLAG_ON(c, EVHTP_CONN_FLAG_KEEPALIVE);
+        /* since the request is keep-alive, assure that the connection
+         * is aware of the same.
+         */
+        HTP_FLAG_ON(conn, EVHTP_CONN_FLAG_KEEPALIVE);
 
-        c->request         = NULL;
-        c->body_bytes_read = 0;
+        conn->body_bytes_read = 0;
 
-        if (c->htp->parent && c->flags & EVHTP_CONN_FLAG_VHOST_VIA_SNI)
+        if (conn->type == evhtp_type_server)
```

Comment:
Would it be appropriate to use evhtp_likely here, or is that pointless?

Reply:
I actually hate the idea that this is a `connection` tagged function - but references other types outside of the `evhtp_connection_t` struct.

I don't think it's needed here since there is not a false condition.